### PR TITLE
Add personal access tokens as an alternative to X-API-* headers

### PR DESCRIPTION
**Pull request by kraih**
A personal access token can be constructed from existing API keys and secrets. The format is `USER:KEY:SECRET`.

This allows for other clients than OpenQA::Client to access the openQA API with authentication for operator and admin privileges. Since the token is submitted via Basic authentication, which pretty much all HTTP clients support, it is as easy as `curl -u kraih:KEY:SECRET ...` to use.

As a security precaution, personal access tokens are currently limited to HTTPS requests originating from any host, and HTTP requests originating from localhost. This will prevent API secrets from being sent in cleartext over the network.

Reverse proxy setups might need to add an `X-Forwarded-Proto: https` header to properly signal to the application server that the connection to the client is actually secure.

Progress: https://progress.opensuse.org/issues/87698

### CODE REVIEW

**Comment by kraih** [comment]
We are reusing the same workflow as the existing API key/secret authentication uses. So only the actual authentication mechanism is exchanged, and code changes are very limited in scope.

I'd really like to add some documentation too, but am not sure where it would best fit in. If nobody has a better suggestion i'll add something to the [Client/Authentication](http://open.qa/docs/#client) section.

File: lib/OpenQA/Shared/Controller/Auth.pm [hunk 1]
```diff
@@ -56,62 +54,82 @@ sub check {
...
+                my ($username, $key, $secret) = ($1, $2, $3);
+                $log->debug(qq{Personal access token for user "$username"});
+                if ($self->is_local_request || $self->req->is_secure) {
+                    my $ip         = $self->tx->remote_address;
+                    my $reject_msg = qq{Rejecting personal access token for user "$username" with ip "$ip"};
+                    if (my $api_key = $self->schema->resultset('ApiKeys')->find({key => $key})) {
+                        my $real_user = $api_key->user;
+                        if ($real_user && secure_compare($real_user->username, $username)) {
+                            if (secure_compare($api_key->secret, $secret)) { ($user, $reason) = ($real_user, undef) }
+                            else                                           { $log->error("$reject_msg, wrong secret") }
```
Comment (Martchus) [inline lib/OpenQA/Shared/Controller/Auth.pm:83]: If a user mis-types its secret, should this really be logged as an error? This is not a problem an admin should be concerned about. The same counts for the other errors.

File: lib/OpenQA/Shared/Controller/Auth.pm [hunk 2]
```diff
@@ -56,62 +54,82 @@ sub check {
     return undef;
 }
 
-sub auth {
-    my $self = shift;
-    my $log  = $self->app->log;
-    my $user;
-
-    my $reason = "Not authorized";
+sub auth ($self) {
```
Comment (Martchus) [inline lib/OpenQA/Shared/Controller/Auth.pm:57]: Maybe it is time to split the code into smaller functions.

**Comment by Martchus** [comment]
About the documentation: I think the section you've found is already a good fit. At least the sentence "And that is usually where you reach the limits of curl" needs to be adjusted anyways.

File: lib/OpenQA/Shared/Controller/Auth.pm (hunk 1 above)
Comment (kraih) [reply lib/OpenQA/Shared/Controller/Auth.pm:83]: I've based this on the API key auth logging. There it's also an error message.

Comment (kraih) [reply lib/OpenQA/Shared/Controller/Auth.pm:83]: Personally, i don't think it's all that useful and would be happy to change it to whatever we can agree on.

File: lib/OpenQA/Shared/Controller/Auth.pm (hunk 2 above)
Comment (kraih) [reply lib/OpenQA/Shared/Controller/Auth.pm:57]: Fair point, i was a bit focused on keeping all the old logic that i ignored opportunities for making the code cleaner.

File: lib/OpenQA/Shared/Controller/Auth.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/Shared/Controller/Auth.pm:83]: I guess it makes sense to adapt the existing API key auth logging then.

File: lib/OpenQA/Shared/Controller/Auth.pm (hunk 2 above)
Comment (Martchus) [reply lib/OpenQA/Shared/Controller/Auth.pm:57]: It also makes sense to do this kind of refactoring as a 2nd step (e.g. 2nd commit or further PR).

**Review by okurz — APPROVED** [review]
This looks great already! I checked the test and there is no unexpected output and runtime is also fine. Of you like, as suggested by Martchus you can show up with some style changes, multiple commits would be appreciated. Also I suggest to reference the ticket in the git commit message

**Review by kalikiana — APPROVED** [review]

**Comment by kraih** [comment]
Ok, i'll make separate PRs for documentation and code cleanup. 😀

**Pull request merged**